### ISSUE TITLE
Supported hooks in "Build your own Trainer" Fabric template don't work

### ISSUE THREAD

**Issue by CSRavasio** [Lightning-AI/lightning#17872]
### Bug description

Hi, I have downloaded the "build your own trainer" files here: https://github.com/Lightning-AI/lightning/tree/master/examples/fabric/build_your_own_trainer

The docstring of the trainer gives a list of hooks that supposedly supported, and they are used that way in the template too. However, none of the hooks work when actually used as methods in the MNISTModule, as tested on the example of `on_train_epoch_start`.

Stepping through, it becomes clear that there are no built-in callbacks available for self.fabric.call(): the first line of call() says `for callback in self._callbacks:`, and `self._callbacks` in this template is an empty list.

In general it will just mean the corresponding methods in the model module are never executed. More specifically though, it also means that this line:

`self.fabric.call("on_validation_model_eval")  # calls 'model.eval()'`

... does not actually do what the comment suggests. `model.training` is `True` before that line, and it remains `True` after it. This might well catch people out who don't use any of the hooks and therefore do not realise this line doesn't have any effect either.

Have I missed something important, is there a way of getting these "supported hooks" to work anyway, or is it indeed a bug and the template should state that all hooks / callbacks must be created first?

### What version are you seeing the problem on?

v2.0

### How to reproduce the bug

```python
Add the following to the MNISTModule class in run.py:

def on_train_epoch_start(self) -> None:
    print("TESTTESTTEST", flush=True)
```

### Error messages and logs

- Silent error: calls are simply not run as the callbacks don't exist internally

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
        - GPU:
                - Quadro RTX 8000
                - Quadro P6000
        - available:         True
        - version:           11.7
* Lightning:
        - lightning:         2.0.3
        - lightning-cloud:   0.5.36
        - lightning-utilities: 0.8.0
        - pytorch-lightning: 2.0.3
        - torch:             2.0.1
        - torchmetrics:      0.11.4
        - torchvision:       0.15.2
* Packages:
        - absl-py:           1.4.0
        - aiohttp:           3.8.3
        - aiosignal:         1.3.1
        - antlr4-python3-runtime: 4.9.3
        - anyio:             3.7.0
        - arrow:             1.2.3
        - async-timeout:     4.0.2
        - attrs:             23.1.0
        - beautifulsoup4:    4.12.2
        - blessed:           1.20.0
        - blinker:           1.6.2
        - cachetools:        4.2.4
        - certifi:           2023.5.7
        - cffi:              1.15.1
        - chardet:           5.1.0
        - charset-normalizer: 2.0.4
        - click:             8.1.3
        - cmake:             3.26.4
        - croniter:          1.3.15
        - cryptography:      39.0.1
        - dateutils:         0.6.12
        - deepdiff:          6.3.0
        - fastapi:           0.88.0
        - faust-cchardet:    2.1.18
        - filelock:          3.12.2
        - frozenlist:        1.3.3
        - fsspec:            2023.6.0
        - google-auth:       2.20.0
        - google-auth-oauthlib: 1.0.0
        - grpcio:            1.48.2
        - h11:               0.14.0
        - h5py:              3.7.0
        - hydra-core:        1.3.2
        - idna:              3.4
        - importlib-metadata: 6.6.0
        - inquirer:          3.1.3
        - itsdangerous:      2.1.2
        - jinja2:            3.1.2
        - lightning:         2.0.3
        - lightning-cloud:   0.5.36
        - lightning-utilities: 0.8.0
        - lit:               16.0.6
        - markdown:          3.4.3
        - markdown-it-py:    3.0.0
        - markupsafe:        2.1.3
        - mdurl:             0.1.2
        - mkl-fft:           1.3.6
        - mkl-random:        1.2.2
        - mkl-service:       2.4.0
        - mpmath:            1.3.0
        - multidict:         6.0.2
        - networkx:          3.1
        - numpy:             1.24.3
        - nvidia-cublas-cu11: 11.10.3.66
        - nvidia-cuda-cupti-cu11: 11.7.101
        - nvidia-cuda-nvrtc-cu11: 11.7.99
        - nvidia-cuda-runtime-cu11: 11.7.99
        - nvidia-cudnn-cu11: 8.5.0.96
        - nvidia-cufft-cu11: 10.9.0.58
        - nvidia-curand-cu11: 10.2.10.91
        - nvidia-cusolver-cu11: 11.4.0.1
        - nvidia-cusparse-cu11: 11.7.4.91
        - nvidia-nccl-cu11:  2.14.3
        - nvidia-nvtx-cu11:  11.7.91
        - oauthlib:          3.2.2
        - omegaconf:         2.3.0
        - opencv-python:     4.7.0.72
        - ordered-set:       4.1.0
        - packaging:         23.1
        - pandas:            2.0.2
        - pillow:            9.5.0
        - pip:               23.1.2
        - protobuf:          3.19.6
        - psutil:            5.9.5
        - pyasn1:            0.4.8
        - pyasn1-modules:    0.2.7
        - pycparser:         2.21
        - pydantic:          1.10.9
        - pygments:          2.15.1
        - pyjwt:             2.7.0
        - pyopenssl:         23.2.0
        - pysocks:           1.7.1
        - python-dateutil:   2.8.2
        - python-editor:     1.0.4
        - python-multipart:  0.0.6
        - pytorch-lightning: 2.0.3
        - pytz:              2023.3
        - pyu2f:             0.1.5
        - pyyaml:            6.0
        - readchar:          4.0.5
        - requests:          2.31.0
        - requests-oauthlib: 1.3.1
        - rich:              13.4.2
        - rsa:               4.9
        - setuptools:        67.8.0
        - six:               1.16.0
        - sniffio:           1.3.0
        - soupsieve:         2.4.1
        - starlette:         0.22.0
        - starsessions:      1.3.0
        - sympy:             1.12
        - tensorboard:       2.13.0
        - tensorboard-data-server: 0.7.1
        - tensorboard-plugin-wit: 1.8.1
        - torch:             2.0.1
        - torchmetrics:      0.11.4
        - torchvision:       0.15.2
        - tqdm:              4.65.0
        - traitlets:         5.9.0
        - triton:            2.0.0
        - typing-extensions: 4.6.3
        - tzdata:            2023.3
        - urllib3:           1.26.16
        - uvicorn:           0.22.0
        - wcwidth:           0.2.6
        - websocket-client:  1.5.3
        - websockets:        11.0.3
        - werkzeug:          2.1.2
        - wheel:             0.38.4
        - yarl:              1.8.1
        - zipp:              3.15.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.11.3
        - release:           5.4.0-147-generic
        - version:           #164~18.04.1-Ubuntu SMP Tue Apr 4 00:20:20 UTC 2023

</details>

cc @carmocca @justusschock @awaelchli

**Comment by CSRavasio** [comment]
Additionally: calling `model.lr_scheduler_step` from `step_scheduler` as the template does causes the following warning:

`[...]/lightning/fabric/wrappers.py:164: PossibleUserWarning: You are calling the method 'MNISTModule.lr_scheduler_step()' from outside the model. This will bypass the wrapper from the strategy and result in incorrect behavior in '.backward()'. You should pass your inputs through '<class '__main__.MNISTModule'>.forward()'.`

I am not sure I understand this warning, and I don't know whether it could be safely ignored in this context, and if so why.

**Comment by CSRavasio** [comment]
Amazing, thank you!

**Comment by pableeto** [comment]
Hi, the current implementation logic would throw an error if -
(1) Using a LightningModule with Fabric;
(2) Not using a ModelHook but instead using a custom callback with same function name as the model hook but with different signatures.

The reason of not using a ModelHook is because it is non-necessary code snippet (e.g. some monitoring, logging, etc). The reason that the callback function has different signatures is because the callback needs to have a reference to the LightningModule or Fabric class.

My current hacky solution would be monkey-patching the ModelHooks to accept arbitrary arguments - is there a better solution that fabric can handle this?

**Comment by LeoLaugier** [comment]
> Additionally: calling `model.lr_scheduler_step` from `step_scheduler` as the template does causes the following warning:
> 
> `[...]/lightning/fabric/wrappers.py:164: PossibleUserWarning: You are calling the method 'MNISTModule.lr_scheduler_step()' from outside the model. This will bypass the wrapper from the strategy and result in incorrect behavior in '.backward()'. You should pass your inputs through '<class '__main__.MNISTModule'>.forward()'.`
> 
> I am not sure I understand this warning, and I don't know whether it could be safely ignored in this context, and if so why.

Hi, I' getting the same warning. Could it be safely ignored? Thanks
